lib/cretonne/src/licm.rs

```diff
@@ -15,40 +20,137 @@ pub fn do_licm(func: &mut Function,
+        let closest_def_loop = closest_def_position(func, inst, inst_loop, loop_analysis);
+        // If all the instruction's argument are defined outside the loop then this instruction is
+        // loop-invariant
+        let destination_loop = match closest_def_loop {
+            // If all the arguments are defined outside any loops we move the instruction just
+            // before the outermost loop containing inst_loop.
+            DefPosition::OutsideLoop() => loop_analysis.outermost_loop(inst_loop, None),
+            // In this case we move just before the loop which would be a direct child of lp
+            DefPosition::ParentLoop(lp) => loop_analysis.outermost_loop(inst_loop, Some(lp)),
+            // If an argument is defined inside the current loop, we abort because the instruction
+            // is not loop-invariant.
+            DefPosition::InsideCurrentLoop() => continue,
+        };
+        match has_pre_header(&func.layout,
+                             cfg,
+                             domtree,
+                             loop_analysis.loop_header(destination_loop)) {
+            None => {
+                let loop_header = loop_analysis.loop_header(destination_loop);
+                let (pre_header, jump_inst) = create_pre_header(loop_header, func, cfg, domtree);
+                loop_analysis.recompute_loop_preheader(pre_header, loop_header);
+                domtree.recompute_loop_preheader(loop_header, pre_header, jump_inst);
+                {
+                    let mut cur = Cursor::new(&mut func.layout);
+                    cur.goto_inst(inst);
+                    cur.remove_inst();
+                    cur.goto_inst(jump_inst);
+                    cur.insert_inst(inst);
+                }
+            }
+            // If there is a natural pre-header we insert new instructions just before the
+            // related jumping instruction (which is not necessarily at the end).
+            Some((_, last_jump_inst)) => {
+                let mut cur = Cursor::new(&mut func.layout);
+                cur.goto_inst(inst);
+                cur.remove_inst();
+                cur.goto_inst(last_jump_inst);
+                cur.insert_inst(inst);
+            }
+        };
+    }
+}
+
+// Given a instruction and the inner-most loop it is part of, determine from the positions of its
+// arguments the closest definition of one of the instruction's argument. This closest position
+// can be:
+// - outside any loop;
+// - inside the `inst_loop`, meaning that the instruction is not loop-invariant;
+// - inside a parent loop which is not `inst_loop`.
+fn closest_def_position(func: &Function,
+                        inst: Inst,
+                        inst_loop: Loop,
+                        loop_analysis: &LoopAnalysis)
+                        -> DefPosition {
+    func.dfg
+        .inst_args(inst)
+        .into_iter()
+        .fold(DefPosition::OutsideLoop(), |acc, arg| {
```

Comment:
I do like expressing this as a fold, in the abstract. However, fold isn't short-circuiting. It seems like we'd want to stop iterating once we find an InsideCurrentLoop operand. I'm assuming it's common to use values defined in the same block, so that terminating this loop early could save a significant amount of work.